packages/Ilcsoftpackage/package.py

```diff
@@ -7,6 +7,69 @@
 
 import os
 
+import llnl.util.tty as tty
+import spack.cmd
+import spack.cmd.common.arguments as arguments
+import spack.environment as ev
+import spack.util.environment
+from  spack.util.environment import *
+import spack.user_environment as uenv
+import spack.store
+import os
+
+
+
+
+
+
+def k4_generate_setup_script(env_mod, shell='sh'):
+    """Return shell code corresponding to a EnvironmentModifications object.
+    Contrary to the spack environment_modifications() method, this does not evaluate
+    the current environment, but generates shell code like:
+    export PATH=/new/path:$PATH
+    instead of:
+    export PATH=/new/path:/current/contents/of/PATH;
+    if `/new/path` is to be prepended.
+
+    :param env_mod: spack EnvironmentModifications object
+    :type env_mod: class: `spack.EnvironmentModifications`
+    :return: Shell code corresponding to the environment modifications.
+    :rtype: str
+    """
+    # first, deduplicate paths # TODO
+    #modifications = self.group_by_name()
+    #for name, actions in sorted(modifications.items()):
+    #  self.prune_duplicate_paths(name)
+    # second, keep track if the paths should be set or prepended
+    modifications = env_mod.group_by_name()
+    new_env = {}
+    env_set_not_prepend = {}
+    for name, actions in sorted(modifications.items()):
+        for x in actions:
+            if isinstance(x, SetPath) or isinstance(x, SetEnv):
+              env_set_not_prepend[name] = True
+            else:
+              env_set_not_prepend[name] = False
+            # third, actually set an environment
+            x.execute(new_env)
+
+    # fourth, get shell commands
+    k4_shell_set_strings = {
+        'sh': 'export {0}={1};\n',
+    }
+    k4_shell_prepend_strings = {
+        'sh': 'export {0}={1}:${0};\n',
+    }
+    cmds = ''
```

Comment:
Can this be made a list to which the commands are appended and then in the end `return ''.join(cmds)`. This might help to avoid having to allocate extremely long strings over and over again.

Reply:
Sure, good point.